File: Wep_Encryption.cpp

#include <iostream>
#include <string>
#include <vector>
#include <cstdint>
#include <iomanip>
#include <bitset>
#include <chrono>
int countCharacters(const std::string& str) {
    return str.size();
}
template<typename Duration, typename Function, typename... Args>
Duration time_taken_by_the_function(const std::string& function_name, Function&& function, Args&&... args) {
    auto Tstart = std::chrono::high_resolution_clock::now();
    std::forward<Function>(function)(std::forward<Args>(args)...);
    auto Tend = std::chrono::high_resolution_clock::now();
    auto time_duration = std::chrono::duration_cast<Duration>(Tend - Tstart);
    std::cout << "Executing the '" << function_name << "' took " << time_duration.count() << " microseconds" << std::endl;
    return time_duration;
}
std::string wep_encrypt(const std::string& plaintext, const std::vector<uint8_t>& wep_key) {
    std::string ciphertext;
    size_t key_length = wep_key.size();
    for (size_t i = 0; i < plaintext.size(); i++) {
        char encrypted_byte = plaintext[i] ^ wep_key[i % key_length];
        ciphertext += encrypted_byte;
    }
    return ciphertext;
}
std::string to_binary(const std::string& input) {
    std::string binary;
    for (const auto& byte : input) {
        binary += std::bitset<8>(byte).to_string();
    }
    return binary;
}
int main() {
    std::string plaintext = "we present how to learn regression models on Lie groups\n"
                            "and apply our formulation to visual object tracking tasks. Many transformations used\n"
                            "order approximation to the geodesic error";
    int Characters_Count =  countCharacters(plaintext);
    std::cout << "Number of characters: " << Characters_Count << std::endl;
    std::vector<uint8_t> wep_key = { 0xAA, 0xBB, 0xCC, 0xDD, 0xEE };
    auto duration = time_taken_by_the_function<std::chrono::microseconds>("wep_encrypt", wep_encrypt, plaintext, wep_key);
    std::string ciphertext = wep_encrypt(plaintext, wep_key);
    return 0;
}

File: Wep_Encryption.rs

//Final Encrypt
use std::time::Instant;
fn count_characters(s: &str) -> usize {
    s.chars().count()
}
fn time_taken_by_the_function<F, R, Args>(function_name: &str, function: F, args: Args) -> R
where
    F: FnOnce(Args) -> R,
{
    let t_start = Instant::now();
    let result = function(args);
    let t_end = Instant::now();
    let time_duration = t_end - t_start;
    println!(
        "Executing the '{}' took {} microseconds",
        function_name,
        time_duration.as_micros()
    );
    result
}
fn wep_encrypt(plaintext_and_key: (&str, &[u8])) -> String {
    let (plaintext, wep_key) = plaintext_and_key;
    let key_length = wep_key.len();
    let mut ciphertext = String::new();
    for (i, byte) in plaintext.bytes().enumerate() {
        let encrypted_byte = byte ^ wep_key[i % key_length];
        ciphertext.push_str(&format!("{:08b}", encrypted_byte));
    }
    ciphertext
}
fn to_binary(input: &str) -> String {
    let mut binary = String::new();
    for byte in input.bytes() {
        binary.push_str(&format!("{:08b}", byte));
    }
    binary
}
fn main() {
    let plaintext = "we present how to learn regression models on Lie groups\n\
                     we present how to learn regression models on Lie groups\n\
                     order approximation to the geodesic error";
    let characters_count = count_characters(plaintext);
    println!("Number of characters: {}", characters_count);
    let wep_key = vec![0xAA, 0xBB, 0xCC, 0xDD, 0xEE];
    let _duration = time_taken_by_the_function("wep_encrypt", wep_encrypt, (plaintext, &wep_key));
    let ciphertext = wep_encrypt((plaintext, &wep_key));
    println!("Ciphertext (binary): {}", ciphertext);
}


